crates/ruma-events/src/message.rs

```diff
@@ -0,0 +1,114 @@
+    pub fn html(body: impl Into<String>) -> Self {
+        Self { mimetype: "text/html".to_owned(), body: body.into() }
+    }
+
+    /// Creates a new HTML-formatted message body by parsing the markdown in `body`.
+    ///
+    /// Returns `None` if no markdown formatting was found.
+    #[cfg(feature = "markdown")]
+    pub fn markdown(body: impl AsRef<str>) -> Option<Self> {
+        let body = body.as_ref();
+        let mut html_body = String::new();
+
+        pulldown_cmark::html::push_html(&mut html_body, pulldown_cmark::Parser::new(body));
+
+        (html_body != format!("<p>{}</p>\n", body)).then(|| Self::html(html_body))
+    }
+
+    fn default_mimetype() -> String {
+        "text/plain".to_owned()
+    }
+}
+
+/// Text message content.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(try_from = "MessageContentSerDeHelper", into = "MessageContentSerDeHelper")]
+pub struct MessageContent(pub(crate) Vec<Text>);
+
+impl MessageContent {
+    /// Get the plain text representation of this message.
+    pub fn plain(&self) -> Option<&str> {
+        self.variants()
+            .iter()
+            .find(|content| content.mimetype == "text/plain")
+            .map(|content| content.body.as_ref())
+    }
+
+    /// Get the HTML representation of this message.
+    pub fn html(&self) -> Option<&str> {
+        self.variants()
+            .iter()
+            .find(|content| content.mimetype == "text/html")
+            .map(|content| content.body.as_ref())
+    }
+
+    /// Get all the text representations of this message.
+    pub fn variants(&self) -> &[Text] {
+        &self.0
+    }
+
+    /// A convenience constructor to create a plain text message.
+    pub fn from_plain(body: impl Into<String>) -> Self {
+        Self(vec![Text::plain(body)])
+    }
+
+    /// A convenience constructor to create an HTML message.
+    pub fn from_html(body: impl Into<String>, html_body: impl Into<String>) -> Self {
+        Self(vec![Text::html(html_body), Text::plain(body)])
+    }
+
+    /// A convenience constructor to create a Markdown message.
+    ///
+    /// Returns an html message if some Markdown formatting was detected, otherwise returns a plain
+    /// text message.
+    #[cfg(feature = "markdown")]
+    pub fn from_markdown(body: impl AsRef<str> + Into<String>) -> Self {
+        let mut message = Vec::with_capacity(2);
+        if let Some(html_body) = Text::markdown(&body) {
+            message.push(html_body);
+        }
+        message.push(Text::plain(body));
+        Self(message)
+    }
+}
+
+/// Helper trait to get the text representation of a message event.
+pub trait TextMessage {
+    /// Get the plain text representation of this message.
+    fn plain(&self) -> Option<&str>;
+
+    /// Get the HTML representation of this message.
+    fn html(&self) -> Option<&str>;
+
+    /// Get all the text representations of this message.
+    fn variants(&self) -> Vec<Text>;
+}
```

Comment:
This trait doesn't make a lot of sense to me, especially because it's not implemented for any types in this module. I also am not sure whether this provides much value in general, so can you leave it out for now?

Reply:
It doesn't make a lot of sense right now because there's only one type that's been "extended", but in the end it should be implemented by all the `m.room.message` events content and the corresponding new types.

Its one of the goals of extended events: if you don't want/know image messages, you can show them as text messages. This trait helps you get the text representation.

But if you prefer I can remove it.

Reply:
Hm, maybe you can split this PR in two? One that just adds the new event types, and one for the `m.room.message` changes which also include that oddly-named new field. Then the first can be merged immediately and the transitional changes can be dealt with separately (I think I'll have to do some reading up on that soon).